### ISSUE TITLE
stream = True not working in video predict for long videos

### ISSUE THREAD

**Issue by bobyfisch** [ultralytics/ultralytics#2930]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

This issue stems from the discussion at https://github.com/ultralytics/ultralytics/issues/1869

1. When trying to predict short (<3min) mp4 or avi videos with a pre-trained model (yolo = YOLO("yolov8n.pt") ) after re-training on a specific dataset, with:
     ```
        yolo.predict(source=self.video_path , project=self.save_path, save=True, save_txt=True, verbose=False, show=False)
     ```
    Prediction works perfectly, saving output videos and labels in self.save_path.

2. When trying to predict longer videos (~10min) the predict function saturates the computer's memory. As suggested by the warning message that appears and by @glenn-jocher [here](https://github.com/ultralytics/ultralytics/issues/1869), stream=True is included in the video predict call:
    ```
    yolo.predict(source=self.video_path , project=self.save_path, save=True, save_txt=True, verbose=False, show=False, stream=True) `
    ```
    This executes without error but leads to no video created in the project=self.save_path folder nor any labels created.

- Also tried with .avi videos, with the same result
- Tried including  video_codec="mp4v", video_preset="ultrafast" as suggested [here](https://github.com/ultralytics/ultralytics/issues/1869) but these are not valid yolo.predict arguments.
- For the time being I have coded functions to split up the video into 1 minute pieces, predict on each of them (with stream=False) and then concatenate the results and labels, which works well but is a bit of an overkill, if anyone is interested in the code I can provide it.

### Environment

- ultralytics 8.0.107, torch 2.0.0 + CUDA 11.7 (NVIDIA GeForce RTX 3080 Ti)
- OS: Ubuntu 20.04.5 LTS
- Python 3.9.12

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@bobyfisch hello! Thank you for reporting this issue. Based on the information provided, your hypothesis that the issue may be caused by a saturation of computer memory due to longer videos seems to be correct. Stream=True is an option designed to address this issue, but there may be certain limitations to its functionality that cause it to fail on your longer videos. While we cannot guarantee that any particular suggestion will resolve the issue, there are a few possible workarounds that may be helpful.

First, we would suggest trying to lower the video's FPS (frames per second), either by manually doing so in preprocessing or by using the FPS argument when using yolo.predict(). By decreasing the number of frames being processed per second, the memory load should also be reduced.

Alternatively, as you have mentioned, you could split up the video into shorter pieces and process each piece separately to avoid the memory issue. This workaround might be a bit overkill, but it is a reliable solution if you're still experiencing the problem.

Thank you once again for your report, and please don't hesitate to ask if you have any further questions or concerns.

**Comment by Laughing-q** [comment]
@bobyfisch hi, `stream=True` would return a python generator and you need a for loop or a `next` to activate it, like this:
```python
for result in model.predict(stream=True):
    pass
```
like the warning msg that appears:
![image](https://github.com/ultralytics/ultralytics/assets/61612323/37a446fe-9b62-4600-8b64-02d278d3a1b7)

**Comment by em-xqm** [comment]
@bobyfisch  How would you still save the label and corresponding image? save_txt=True and save=True don't seem to work  with the stream=True param?

**Comment by Laughing-q** [comment]
@em-xqm it works with me, but using a for loop to activate the generator.
```python
from ultralytics import YOLO

model = YOLO("yolov8n.pt")
for r in model.predict(source="/home/laughing/Videos/test.mp4", save=True, stream=True, save_txt=True):
    pass
```
and how's what my folder looks like:
![image](https://github.com/ultralytics/ultralytics/assets/61612323/24dd3796-b03e-4e8a-9cd4-7ed5561c1fea)
you can see there are `labels/` and saved `test.mp4`.
